### DISCUSSION TITLE
Filtering IP Addresses results in only a few matches? Can't get all matches?

### DISCUSSION THREAD

**Discussion by bluikko** [netbox-community/netbox#8384] [Q&A] [answered]
While trying to clarify a situation at hand with one server that has multiple IP addresses assigned to it, I came across behavior that looks very strange to me. And I could not find a solution so I have to ask if a) this behavior is as intended and b) if someone has any suggestions how to make a search that I need.

Amongst other servers I have 2 machines, say `x-srv001` and `x-srv002`. They both have 2 NICs connected with IPv4 + IPv6 addresses assigned.  
On both servers one of the interfaces has about 10 IPv4 addresses assigned.  
_What I am attempting to do is list all IP addresses assigned to those 2 servers._

So the natural way would seem to be to browse to `/ipam/ip-addresses/` and try the Quick Search.  
I type `x-srv0` in Quick Search that will match just those 2 servers I am interested in, using the server name.  
_The result:_ 1 IPv6 address from each server. Wait, what? Where are the other IPv6 addresses and all the IPv4 addresses?

Let's try that again, maybe a Filter will work and I will get all the IP addresses. In all the years with NB I've never used it before.  
Wait, there's no query fields for "Interface Name" or "Host Name"? Well, maybe the Search field could work. So I put `x-srv0` in it.  
_The result:_ 1 IPv6 address from each server again.

So it would seem that there is no way to display all IP addresses from hosts matching a query? Wouldn't this be a major missing feature?  
Are both the Filter and Quick Search functions supposed to return only 1 hit per host?

I have not looked at GraphQL, maybe that could work better? But it also looks like quite a heavy solution, I will learn it if necessary and it could accomplish "proper" searches.

**Comment by bluikko** [comment]
And it gets even more strange.

Say I have about 20 different hosts, let's call them `pref-a-srv00[1-9]`, `pref-b-srv00[1-7]` and `pref-db-srv00[1-5]`.  
Viewing `/ipam/ip-addresses/` and hitting browser search function, typing `pref` I get 100 hits, meaning on those about 20 different hosts there are 100 IPv4/IPv6 addresses assigned (`pref` is only used in the host name of those hosts).

Doing a Quick Search for `pref` results in only 6 IPv6 addresses, one from each host! Where are all the ~15 other hosts?

The Filter function gives exactly the same results.

If I use the search box at the top of each page, type in `pref` and set Object Type to `IP Addresses`, I get the exact same 6 results?!  
Is there really no way to:
- return all hosts that match the query; and
- return all IP addresses for each host that match the query?

**Comment by candlerb** [comment, marked as answer]
This is only a guess, but have you put the hostname in the *description* or *DNS name* fields of one of the IP addresses?  That's what the IP address quick search [filters on](https://github.com/netbox-community/netbox/blob/v3.1.6/netbox/ipam/filtersets.py#L589-L593).

In the GUI filters, currently you can only search on "Assigned to an interface (yes/no)".  However, the API already does let you filter on [device or virtual machine](https://github.com/netbox-community/netbox/blob/v3.1.6/netbox/ipam/filtersets.py#L527-L546).  Therefore, I think it would be a simple extension to make this available in the GUI, and you could raise this as a feature request.

In the mean time, you can construct this URL manually in the GUI: `/ipam/ip-addresses/?device_id=1` (replacing "1" with the relevant device ID).  It works for me, in Netbox 3.1.6.

**Reply by bluikko** [reply]
Thank you. I think you had the right idea - "DNS name" field is set and I guess that is the problem then. The DNS name is set only on the "primary" IPv6 address.

I could accomplish the task on hand with `?device_id=1`, didn't think about that. I use the API (and API docs) all the time for automation so I am familiar with that - are all the query parameters available in the API also usable in the web UI by manually appending in the query string? I guess I should test instead of asking...

**Reply by candlerb** [reply]
> are all the query parameters available in the API also usable in the web UI by manually appending in the query string?

I linked to the FilterSet object, and as far as I can see, the same FilterSet is used by both the REST API and the GUI [view](https://github.com/netbox-community/netbox/blob/v3.1.6/netbox/ipam/views.py#L618-L622).

Some other things you could try:
* Add a [Custom Link](https://netbox.readthedocs.io/en/stable/models/extras/customlink/) on Device, which links to `/ipam/ip-addresses/?device_id=<N>`
* Write an [Export Template](https://netbox.readthedocs.io/en/stable/customization/export-templates/) for Device, which lists all the IP addresses on that device (iterate over interfaces and IP addresses)

**Comment by jeremystretch** [comment]
> What I am attempting to do is list all IP addresses assigned to those 2 servers.

`GET /ipam/ip-addresses/?device=x-srv001&device=x-srv002`
